python/pyiceberg/utils/config.py

```diff
@@ -0,0 +1,125 @@
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+import logging
+import os
+import site
+import sys
+from typing import Any, List
+
+import yaml
+
+from pyiceberg.typedef import EMPTY_DICT, Configuration, FrozenDict
+
+PYICEBERG = "pyiceberg__"
+CATALOG = "catalog"
+
+PATHS = [
+    f"{os.environ.get('ICEBERG_ROOT_CONFIG', '/etc/pyiceberg').rstrip('/')}/pyiceberg",
+    *[f"{prefix}/etc/pyiceberg" for prefix in site.PREFIXES],
+    f"{sys.prefix}/etc/pyiceberg",
+    "~/config/pyiceberg",
+]
+
+
+def merge_dicts(lhs: Configuration, rhs: Configuration) -> Configuration:
+    """merges left-hand side into the right-hand side"""
+    for lhs_key, lhs_value in lhs.items():
+        if lhs_key in rhs:
+            rhs_value = rhs[lhs_key]
+            if isinstance(lhs_value, dict) and isinstance(rhs_value, dict):
+                lhs[lhs_key] = merge_dicts(lhs_value, rhs_value)
+            elif lhs_value != lhs_value:
+                lhs[lhs_key] = rhs_value
+
+    # Add the keys from the rhs that are not in the lhs
+    for rhs_key, rhs_value in rhs.items():
+        if rhs_key not in lhs:
+            lhs[rhs_key] = rhs_value
+
+    return lhs
+
+
+logger = logging.getLogger(__name__)
+
+
+def _lowercase_dictionary_keys(input_dict: Configuration) -> Configuration:
+    """Lowers all the keys of a dictionary in a recursive manner, to make the lookup case-insensitive"""
+    return {k.lower(): _lowercase_dictionary_keys(v) if isinstance(v, dict) else v for k, v in input_dict.items()}
+
+
+class Config:
+    config: Configuration
+
+    def __init__(self):
+        config = self._from_configuration_files({})
+        config = merge_dicts(config, self._from_environment_variables(config))
+        self.config = FrozenDict(**config)
+
+    @staticmethod
+    def _from_configuration_files(config: Configuration) -> Configuration:
+        """Traverses directories searching for configuration files
+
+        Takes preference over configuration that's close to the user
+        """
+
+        for path in PATHS:
+            if os.path.isfile(path):
+                with open(path, encoding="utf-8") as f:
+                    try:
+                        file_config = yaml.safe_load(f)
+                        file_config_lowercase = _lowercase_dictionary_keys(file_config)
+                        config = merge_dicts(config, file_config_lowercase)
+                    except yaml.YAMLError:
+                        logger.exception("Not valid config yaml: %s", path)
+        return config
+
+    @staticmethod
+    def _from_environment_variables(config: Configuration) -> Configuration:
```

Comment:
I think this is too complicated. If we want to support environment variables, we should make them independent and configure only a single catalog. I don't think there's a use case for configuring multiple catalogs through a ton of environment variables.

I think we should carefully define what environment variables do and have a static set of them. I like `PYICEBERG_URI` and `PYICEBERG_CREDENTIAL`. I think those should set up a default catalog. If you don't specify a catalog, those are used. If you do specify a catalog, we should ignore them. That way we don't need to worry about conflicts, merging, or all this complexity of parsing environment variable names.

Reply:
I do like this a lot. I got this idea from Dask as well: https://docs.dask.org/en/stable/configuration.html#environment-variables. But also tool like Airflow rely heavily on this: https://airflow.apache.org/docs/apache-airflow/stable/howto/set-config.html This is not only for configuring Airflow itself, but also the connections: https://airflow.apache.org/docs/apache-airflow/stable/howto/connection.html#storing-connections-in-environment-variables

When the `pyiceberg` CLI gets used in orchestrators such as Airflow or even as a k8s cron, environment variables are often a convenient and powerful way of setting the credentials. 

> I think those should set up a default catalog. If you don't specify a catalog, those are used. If you do specify a catalog, we should ignore them

I agree with that, I'll take that into consideration in the CLI PR.

Reply:
If Dask and Airflow both have this convention then there's sufficient precedent. I'd probably add it later, but it should be fine.